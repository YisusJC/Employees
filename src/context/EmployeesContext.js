import { createContext, useContext } from "react";

export const AppContext = createContext();
export const useEmployeesContext = () => useContext(AppContext);
export default AppContext;
