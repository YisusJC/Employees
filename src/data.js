const empData = [
  {
    name: "Jaime Armando Straus",
    address: "Los Angeles, CA",
    dni: "MBJ34368V12P5",
    birth: "19/03/1998",
    phone: "5628903455",
    email: "jaime.armando@aspiresys.com",
    position: "Front-end Developer",
  },
  {
    name: "Jose Armando Straus",
    address: "Washington, DC",
    dni: "NFIN3543543NKBN",
    birth: "19/03/1980",
    phone: "5628903455",
    email: "jose.armando@aspiresys.com",
    position: "Back-end Developer",
  },
];

export const employees = [...empData];
export const setEmployees = jest.fn();
