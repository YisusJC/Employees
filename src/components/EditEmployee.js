import React, { useContext, useState } from 'react';
import unid from 'uniqid';
import { EmployeesContext } from '../context/EmployeesContext';
import { useNavigate, useParams } from 'react-router-dom'; 

export const EditEmployee = ({id = "MBJ34368V12P5"}) => {
    const navigate = useNavigate();
    const params = useParams();

    id = params.id;

    const {employees, setEmployees} = useContext(EmployeesContext)

    const [employee, setEmployee] = useState(employees.find(({dni}) => dni === id));

    const saveEmployee = (id, emp) => {
        setEmployees(employees.map(employee => employee.dni === id ? employee = emp : employee));
        navigate("/")
    }
    
    return (
        <div className='add-employee'>
            <div className="field">
                <label>Name</label>
                <input type="text" placeholder='Full Name' value={employee.name} onInput={e => setEmployee({...employee, name: e.target.value})} />
            </div>
            <div className="field">
                <label>DNI</label>
                <input type="text" placeholder='DNI' value={employee.dni} onInput={e => setEmployee({...employee, dni: e.target.value})} />
                <button onClick={() => {setEmployee({...employee, dni: ""}); setEmployee({...employee, dni: unid().toUpperCase()})}}>Generate</button>
            </div>
            <div className="field">
                <label>Birth date</label>
                <input type="text" placeholder='DD/MM/YYYY' value={employee.birth} onInput={e => setEmployee({...employee, birth: e.target.value})} />
            </div>
            <div className="field">
                <label>Address</label>
                <input type="text" placeholder='Address' value={employee.address} onInput={e => setEmployee({...employee, address: e.target.value})} />
            </div>
            <div className="field">
                <label>Phone</label>
                <input type="text" placeholder='Phone' value={employee.phone} onInput={e => setEmployee({...employee, phone: e.target.value})} />
            </div>
            <div className="field">
                <label>Email</label>
                <input type="text" placeholder='Email' value={employee.email} onInput={e => setEmployee({...employee, email: e.target.value})} />
            </div>
            <div className="field">
                <label>Position</label>
                <input type="text" placeholder='Position' value={employee.position} onInput={e => setEmployee({...employee, position: e.target.value})} />
            </div>

            <div className="actions">
                <button onClick={() => saveEmployee(id, employee)}>Confirm</button>
                <button onClick={() => navigate("/")}>Cancel</button>
            </div>
        </div>
    )
}
