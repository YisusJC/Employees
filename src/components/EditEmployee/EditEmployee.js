import React, { useState } from "react";
import unid from "uniqid";
import { useEmployeesContext } from "../../context/EmployeesContext";
import { useNavigate, useParams } from "react-router-dom";

export const EditEmployee = () => {
  const navigate = useNavigate();
  const params = useParams();

  const id = params.id;

  const { employees, setEmployees } = useEmployeesContext();

  const [employee, setEmployee] = useState(
    employees.find(({ dni }) => dni === id)
  );

  const saveEmployee = (id, emp) => {
    setEmployees(
      employees.map((employee) =>
        employee.dni === id ? (employee = emp) : employee
      )
    );
    navigate("/");
  };

  return (
    <div className="add-employee">
      <div className="field">
        <label>Name</label>
        <input
          type="text"
          placeholder="Full Name"
          id="name"
          value={employee.name}
          onChange={(e) => setEmployee({ ...employee, name: e.target.value })}
        />
      </div>
      <div className="field">
        <label>DNI</label>
        <input
          type="text"
          placeholder="DNI"
          id="dni"
          value={employee.dni}
          onChange={(e) => setEmployee({ ...employee, dni: e.target.value })}
        />
        <button
          id="generate-dni"
          onClick={() => {
            setEmployee({ ...employee, dni: "" });
            setEmployee({ ...employee, dni: unid().toUpperCase() });
          }}
        >
          Generate
        </button>
      </div>
      <div className="field">
        <label>Birth date</label>
        <input
          type="text"
          placeholder="DD/MM/YYYY"
          id="birth"
          value={employee.birth}
          onChange={(e) => setEmployee({ ...employee, birth: e.target.value })}
        />
      </div>
      <div className="field">
        <label>Address</label>
        <input
          type="text"
          placeholder="Address"
          id="address"
          value={employee.address}
          onChange={(e) =>
            setEmployee({ ...employee, address: e.target.value })
          }
        />
      </div>
      <div className="field">
        <label>Phone</label>
        <input
          type="text"
          placeholder="Phone"
          id="phone"
          value={employee.phone}
          onChange={(e) => setEmployee({ ...employee, phone: e.target.value })}
        />
      </div>
      <div className="field">
        <label>Email</label>
        <input
          type="text"
          placeholder="Email"
          id="email"
          value={employee.email}
          onChange={(e) => setEmployee({ ...employee, email: e.target.value })}
        />
      </div>
      <div className="field">
        <label>Position</label>
        <input
          type="text"
          id="position"
          placeholder="Position"
          value={employee.position}
          onChange={(e) =>
            setEmployee({ ...employee, position: e.target.value })
          }
        />
      </div>

      <div className="actions">
        <button id="save-employee" onClick={() => saveEmployee(id, employee)}>
          Confirm
        </button>
        <button id="cancel-employee" onClick={() => navigate("/")}>
          Cancel
        </button>
      </div>
    </div>
  );
};
