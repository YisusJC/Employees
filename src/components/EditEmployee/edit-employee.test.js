import { EditEmployee } from "./EditEmployee";
import { mount, shallow } from "enzyme/build";
import { AppContext } from "../../context/EmployeesContext";
import { employees, setEmployees } from "../../data";
import { BrowserRouter as Router } from "react-router-dom";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({
    id: "MBJ34368V12P5",
  }),
  useRouteMatch: () => ({ url: "/edit-employee/MBJ34368V12P5" }),
}));

// Add employee Jest tests
describe("Add employee Component", () => {
  // should render without crashing.
  it("should render without crashing", () => {
    const wrapper = shallow(
      <AppContext.Provider value={{ employees, setEmployees }}>
        <EditEmployee />
      </AppContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  // should execute saveEmployee function when click button.
  it("should execute saveEmployee function when click button", () => {
    const wrapper = mount(
      <AppContext.Provider value={{ employees, setEmployees }}>
        <Router>
          <EditEmployee />
        </Router>
      </AppContext.Provider>
    );
    const editEmployee = wrapper.find(EditEmployee).first();
    const button = editEmployee.find("#save-employee");
    button.simulate("click");
    expect(button.props().onClick).toBeDefined();
  });

  // should navigate to / in case cancel save employee.
  it("should navigate to / in case cancel save employee", () => {
    const wrapper = mount(
      <AppContext.Provider value={{ employees, setEmployees }}>
        <Router>
          <EditEmployee />
        </Router>
      </AppContext.Provider>
    );
    const editEmployee = wrapper.find(EditEmployee).first();
    const button = editEmployee.find("#cancel-employee");
    button.simulate("click");
    expect(button.props().onClick).toBeDefined();
  });

  // should generate DNI in case click button and save dni with setEmployee.
  it("should generate DNI in case click button", () => {
    const wrapper = mount(
      <AppContext.Provider value={{ employees, setEmployees }}>
        <Router>
          <EditEmployee />
        </Router>
      </AppContext.Provider>
    );
    const editEmployee = wrapper.find(EditEmployee).first();
    const button = editEmployee.find("#generate-dni");
    button.simulate("click");
    expect(button.props().onClick).toBeDefined();
  });

  // should render and save input name
  it("should render and save input name", () => {
    const wrapper = mount(
      <AppContext.Provider value={{ employees, setEmployees }}>
        <Router>
          <EditEmployee />
        </Router>
      </AppContext.Provider>
    );
    const editEmployee = wrapper.find(EditEmployee).first();
    const input = editEmployee.find("#name");
    expect(input.props().value).toBe("Jaime Armando Straus");
  });

  // should render and save input dni
  it("should render and save input dni", () => {
    const wrapper = mount(
      <AppContext.Provider value={{ employees, setEmployees }}>
        <Router>
          <EditEmployee />
        </Router>
      </AppContext.Provider>
    );
    const editEmployee = wrapper.find(EditEmployee).first();
    const input = editEmployee.find("#dni");
    expect(input.props().value).toBe("MBJ34368V12P5");
  });

  // should render and save input birth
  it("should render and save input birth", () => {
    const wrapper = mount(
      <AppContext.Provider value={{ employees, setEmployees }}>
        <Router>
          <EditEmployee />
        </Router>
      </AppContext.Provider>
    );
    const editEmployee = wrapper.find(EditEmployee).first();
    const input = editEmployee.find("#birth");
    expect(input.props().value).toBe("19/03/1998");
  });

  // should render and save input address
  it("should render and save input address", () => {
    const wrapper = mount(
      <AppContext.Provider value={{ employees, setEmployees }}>
        <Router>
          <EditEmployee />
        </Router>
      </AppContext.Provider>
    );
    const editEmployee = wrapper.find(EditEmployee).first();
    const input = editEmployee.find("#address");
    expect(input.props().value).toBe("Los Angeles, CA");
  });

  // should render and save input phone
  it("should render and save input phone", () => {
    const wrapper = mount(
      <AppContext.Provider value={{ employees, setEmployees }}>
        <Router>
          <EditEmployee />
        </Router>
      </AppContext.Provider>
    );
    const editEmployee = wrapper.find(EditEmployee).first();
    const input = editEmployee.find("#phone");
    expect(input.props().value).toBe("5628903455");
  });

  // should render and save input email
  it("should render and save input email", () => {
    const wrapper = mount(
      <AppContext.Provider value={{ employees, setEmployees }}>
        <Router>
          <EditEmployee />
        </Router>
      </AppContext.Provider>
    );
    const editEmployee = wrapper.find(EditEmployee).first();
    const input = editEmployee.find("#email");
    expect(input.props().value).toBe("jaime.armando@aspiresys.com");
  });

  // should render and save input position
  it("should render and save input position", () => {
    const wrapper = mount(
      <AppContext.Provider value={{ employees, setEmployees }}>
        <Router>
          <EditEmployee />
        </Router>
      </AppContext.Provider>
    );

    const editEmployee = wrapper.find(EditEmployee).first();
    const input = editEmployee.find("#position");
    expect(input.props().value).toBe("Front-end Developer");
  });

  // should render and save input name
  it("should render and save input name", () => {
    const wrapper = mount(
      <AppContext.Provider value={{ employees, setEmployees }}>
        <Router>
          <EditEmployee />
        </Router>
      </AppContext.Provider>
    );
    const editEmployee = wrapper.find(EditEmployee).first();
    let input = editEmployee.find("#name");
    input.simulate("change", {
      target: {
        value: "Jaime Armando Straus",
      },
    });
    input = wrapper.find("#name");
    expect(input.props().value).toBe("Jaime Armando Straus");
  });

  // should render and save input dni
  it("should render and save input dni", () => {
    const wrapper = mount(
      <AppContext.Provider value={{ employees, setEmployees }}>
        <Router>
          <EditEmployee />
        </Router>
      </AppContext.Provider>
    );
    const editEmployee = wrapper.find(EditEmployee).first();
    let input = editEmployee.find("#dni");
    input.simulate("change", {
      target: {
        value: "MBJ34368V12P5",
      },
    });
    input = wrapper.find("#dni");
    expect(input.props().value).toBe("MBJ34368V12P5");
  });

  // should render and save input birth
  it("should render and save input birth", () => {
    const wrapper = mount(
      <AppContext.Provider value={{ employees, setEmployees }}>
        <Router>
          <EditEmployee />
        </Router>
      </AppContext.Provider>
    );
    const editEmployee = wrapper.find(EditEmployee).first();
    let input = editEmployee.find("#birth");
    input.simulate("change", {
      target: {
        value: "19/03/1998",
      },
    });
    input = wrapper.find("#birth");
    expect(input.props().value).toBe("19/03/1998");
  });

  // should render and save input address
  it("should render and save input address", () => {
    const wrapper = mount(
      <AppContext.Provider value={{ employees, setEmployees }}>
        <Router>
          <EditEmployee />
        </Router>
      </AppContext.Provider>
    );
    const editEmployee = wrapper.find(EditEmployee).first();
    let input = editEmployee.find("#address");
    input.simulate("change", {
      target: {
        value: "Los Angeles, CA",
      },
    });
    input = wrapper.find("#address");
    expect(input.props().value).toBe("Los Angeles, CA");
  });

  // should render and save input phone
  it("should render and save input phone", () => {
    const wrapper = mount(
      <AppContext.Provider value={{ employees, setEmployees }}>
        <Router>
          <EditEmployee />
        </Router>
      </AppContext.Provider>
    );
    const editEmployee = wrapper.find(EditEmployee).first();
    let input = editEmployee.find("#phone");
    input.simulate("change", {
      target: {
        value: "5628903455",
      },
    });
    input = wrapper.find("#phone");
    expect(input.props().value).toBe("5628903455");
  });

  // should render and save input email
  it("should render and save input email", () => {
    const wrapper = mount(
      <AppContext.Provider value={{ employees, setEmployees }}>
        <Router>
          <EditEmployee />
        </Router>
      </AppContext.Provider>
    );
    const editEmployee = wrapper.find(EditEmployee).first();
    let input = editEmployee.find("#email");
    input.simulate("change", {
      target: {
        value: "jaime.armando@aspiresys.com",
      },
    });
    input = wrapper.find("#email");
    expect(input.props().value).toBe("jaime.armando@aspiresys.com");
  });

  // should render and save input position
  it("should render and save input position", () => {
    const wrapper = mount(
      <AppContext.Provider value={{ employees, setEmployees }}>
        <Router>
          <EditEmployee />
        </Router>
      </AppContext.Provider>
    );

    const editEmployee = wrapper.find(EditEmployee).first();
    let input = editEmployee.find("#position");
    input.simulate("change", {
      target: {
        value: "Front-end Developer",
      },
    });
    input = wrapper.find("#position");
    expect(input.props().value).toBe("Front-end Developer");
  });
});
