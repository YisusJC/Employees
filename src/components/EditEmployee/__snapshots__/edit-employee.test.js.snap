// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add employee Component should render without crashing 1`] = `ShallowWrapper {}`;
